### ISSUE TITLE
Booting no longer works on an Apple IIe

### ISSUE THREAD

**Issue by bokowski** [freitz85/AppleIISd#3]
Booting the example image from an SD card does not work when using the latest state from the master branch. I ordered PCBs using the current Gerber files, configured the CLPD and burnt an EPROM and couldn't get it to work in my Apple IIe. It did work after I decided to try burning the EPROM image from commit f199354 ("ProDOS image for testing added").
I'm happy now but thought it would be useful to report this problem to you.

**Comment by freitz85** [comment]
Thank you for letting me know, I will look into it!
Between those two commits, the Smartport driver was added and since nothing has changed.
Just to make sure, are you using an enhanced IIe with a 65C02 CPU?

**Comment by bokowski** [comment]
Yes, I have an enhanced IIe with the 65C02 CPU. By the way, I completely
forgot to thank you for publishing this project, great work!

**Comment by rapiqui** [comment]
Hi,

I am experiencing the same behavior commented by @bokowski .
I do have an enhanced IIe with a WDC65C02 (W65C02S6TPG-14 with pin 8 and 36 connected as per WDC note) and CD, EF and Video ROMS enhanced.
When I do try to boot a ProDOS image copied to a 2Gb SD with CiderPress I always get prompted to the System Monitor:

![IMG_20200811_222424](https://user-images.githubusercontent.com/55879300/89950157-53727000-dc29-11ea-92fb-bf4a3cf66ff0.jpg)

**Comment by freitz85** [comment]
@rapiqui I have just pushed a change in commit 758abfde4b54e888f47e320f820b1ec8bff0dc5f that fixed an issue that the firmware could overwrite itself. Are you able to build the firmware or do you need a binary image?

In the fixes branch, I am still working on problems in the smartport part of the driver. Pure ProDOS mode (2 partitions) seems to work better there.

**Comment by rapiqui** [comment]
Hi @freitz85 .
First of all big THANKS for this amazing project ;)

You would make me things easier if you could upload as well the binary image so I could update de EPROM content with the programmer.

If it can be helpful: just to mention that I have tried to boot the SD on the the Apple IIe slots 7,6,2 and 1 with the same results.
Tried your ProFile.2mg image and the latest version of Total Replay (4.0 alpha 2) with the same SysMon address shown.
Tried too partitioning the SD with CiderPress selecting a single ProDisk partition and 4/6 partitions which made no difference at all.

Update: I have compiled the firmware and uploaded on the EPROM.
Same behavior, now showing address 803:

![IMG_20200812_163053](https://user-images.githubusercontent.com/55879300/90028181-a431a980-dcb9-11ea-8ccc-4d1f04614170.jpg)

**Comment by freitz85** [comment]
@rapiqui On startup, after initializing the SD card, the ProDOS bootloader is loaded from the first two blocks on the card to address $800 and then is called by a jump to $801, so $803 is the second byte after the start of it. When you are in this error, could you please type `800.880` or `801L` and post the output. This should show if the bootsector has been loaded correctly from the card.

Does the AppleIISd work when booted from a floppy, i.e. can you access any files? If yes, the problem might be in a corrupted image.

Have you followed my [posting ](http://bluemeanie-retro.blogspot.com/2017/12/how-to-prepare-sd-cards-for-appleiisd.html) on how to prepare an SD card? Under certain circumstances, Ciderpress may create a partition that is not on the physical start of the card and the AppleIISd can't handle this.

**Comment by rapiqui** [comment]
@freitz85 Thanks for checking it out.
Yes, I did follow your post regarding CiderPress usage to create an usable AppleIISD SD Card.
Eventually I even had to switch from a 16Gb SD to an old 2Gb and afterwards to a 8Gb as CiderPress stated that volumes larger than 8 Gb weren't supported.

In regards of the floppy question: unfortunately I do not have one. I do have though the floppy disk controller.
I have been trying to boot AppleIISD without the FDC, as doing with it will just remain stuck.

I have took a picture of the requested memory dump.
Please, let me know if you need additional info.

![IMG_20200819_151315](https://user-images.githubusercontent.com/55879300/90639593-288eaa00-e22f-11ea-9b25-ed79ee9f8e96.jpg)

Cheers

**Comment by freitz85** [comment]
@rapiqui This looks like either the bootloader has not been copied at all, or that the bootloader has done its first part and then got stuck somewhere else. The first possibility doesn't fit entirely, because then the address should have been 801 and not 803.
If you have a serial card, you could try booting via ADTPro to get a ProDOS onto the machine and see if the ApleIISd is working at all. 

I have had a couple of reports that some SD cards are not working and I still have to find out why. I am working on a debug build that is using a serial card to put out status info on where it is in the init and boot process. I will keep you posted.

**Comment by rapiqui** [comment]
Thanks @freitz85 I had my suspects on the tested SD cards format, and re-formatted them again using the "SD Card Formatter" v.5.0.1 from Tuxera, which is the recommended tool by the SD consortium, but had no luck at all. All the 3 SD cards tested (2 Gb non-HC or XC), 8 Gb SDHC and 16 Gb HC behave the same way.
I will keep my fingers crossed and hope you could find out the reason why this could be happening :)

In regards of the serial card, unfortunately I do not have one, but happy to help and try to find why. If I need one in order that I could provide you more useful information I can try to find one.

**Comment by freitz85** [comment]
It admittedly took a very long time, but I managed to push some fixes in b13ed9077f54387077b36edc959048bec4b9f689. Please try if this fixed the booting issues as well as the problems with some SD cards.

**Comment by rapiqui** [comment]
Hi,

I have burned the new binary on the EPROM and tested with the //e.
No luck. Still appears stuck at 0803. Strangely this time the $800 - $880 appears blank

![IMG_20210825_163757](https://user-images.githubusercontent.com/55879300/130812042-a876b89a-8df2-4545-a11b-ee9d161c603c.jpg)

**Comment by timmcross** [comment]
I am also seeing unstable behavior on my enhanced, platinum Apple IIe (2 memory chips).  It seems to run fine on my IIgs but on the IIe, it takes a few reboots before the card will boot properly.  Other times, it breaks to the monitor.  I can get screencaps if you like.  I will also try it on my older enhanced IIe (8 memory chips).

**Comment by freitz85** [comment]
I have had the feedback from some users that they had similar issues but the reason seemed to be with the 74LS245. After swapping it worked, specifically on an Apple II+ with the firmware from the 6502 branch. Older LS parts seem to work better than newer ones, for whatever reason.
Unfortunately I am not able to work on the project right now, but what could be done is to compare the VHDL simulation that is already present in the project with a sample of the card working in an actual computer using a logic analyzer. According to Apple's tech pubs, both a HCT and LS should work, but I have had troubles with HCT as well.

**Comment by timmcross** [comment]
Ah, yes.  That seemed to be the issue.  I was using a chip labeled as a 74LS245 I got from Amazon and had issues.  I pulled a 74LS245 from another Apple II motherboard and that worked much better.  Perhaps my Amazon 245 is actually a 74HC245.  Thanks!

**Comment by rittwage-zz** [comment]
I have the opposite problem. Mine is not stable on a IIGS either...

I have a ROM1 GS with a 4M RAM card (although that part I don't believe matters- I've tried the Apple 1.5M card also) and the card is in Slot 7. When I boot GSOS from the SD card, usually some time late in the load process (not usually always the same place) I will hear a "bonk" (IIgs beep) and it will hang. I can reboot with the keyboard reset, and eventually it will boot fine and run stable, sometimes after a couple of tries. I've replace the SD card and this doesn't seem to matter. 

Is there a small issue with the driver, or is it my IIgs?

I can order some other 74LS245 if that is thought to help...

**Comment by Dirk-** [comment]
> I have a ROM1 GS with a 4M RAM card (although that part I don't believe matters- I've tried the Apple 1.5M card also) and the card is in Slot 7. When I boot GSOS from the SD card, usually some time late in the load process (not usually always the same place) I will hear a "bonk" (IIgs beep) and it will hang. I can reboot with the keyboard reset, and eventually it will boot fine and run stable, sometimes after a couple of tries. I've replace the SD card and this doesn't seem to matter.

@rittwage Very similar to my problem. I have a ROM03 GS and the card sometimes hangs during boot, sometimes when launching programs. It is not reliable.

My card came fully equipped from Jens Fischer (who handles the "[shop](http://bluemeanie-retro.blogspot.com/2020/04/prebuilt-appleiisd-v121-finally.html)" for @freitz85, I think). I updated the firmware to the latest version, but this didn't help. I wouldn't mind buying another 74LS245 if someone knew a reliable source.

**Comment by rittwage-zz** [comment]
I tried a different brand (and older) 74LS245 as well as a newer 74HCT245. The behavior did not change, unfortunately.
